File: src/api/tools.ts
import { request } from "./http";
import type {
  AnnuaireDebugResult,
  SireneNewBusiness,
  SireneNewBusinessDirector,
  SireneNewBusinessesResult,
} from "../types";

export interface SireneNewBusinessesPayload {
  startDate: string;
  endDate?: string;
  nafCodes: string[];
  limit?: number;
  departmentCodes?: string[];
  enrichAnnuaire?: boolean;
}

interface SireneNewBusinessDirectorResponse {
  type_dirigeant: string;
  first_names: string | null;
  last_name: string | null;
  quality: string | null;
  birth_month: number | null;
  birth_year: number | null;
  siren: string | null;
  denomination: string | null;
  nationality: string | null;
}

interface SireneNewBusinessResponse {
  siret: string;
  siren: string | null;
  nic: string | null;
  name: string | null;
  naf_code: string | null;
  naf_label: string | null;
  date_creation: string | null;
  is_individual: boolean;
  leader_name: string | null;
  denomination_unite_legale: string | null;
  denomination_usuelle_unite_legale: string | null;
  denomination_usuelle_etablissement: string | null;
  enseigne1: string | null;
  enseigne2: string | null;
  enseigne3: string | null;
  complement_adresse: string | null;
  numero_voie: string | null;
  indice_repetition: string | null;
  type_voie: string | null;
  libelle_voie: string | null;
  code_postal: string | null;
  libelle_commune: string | null;
  libelle_commune_etranger: string | null;
  legal_unit_name: string | null;
  directors: SireneNewBusinessDirectorResponse[];
}

interface SireneNewBusinessesResponse {
  total: number;
  returned: number;
  establishments: SireneNewBusinessResponse[];
}

interface AnnuaireDebugResponse {
  siret: string;
  siren: string;
  success: boolean;
  status_code: number | null;
  duration_ms: number | null;
  error: string | null;
  payload: Record<string, unknown> | null;
}

const mapSireneDirector = (payload: SireneNewBusinessDirectorResponse): SireneNewBusinessDirector => ({
  typeDirigeant: payload.type_dirigeant,
  firstNames: payload.first_names ?? null,
  lastName: payload.last_name ?? null,
  quality: payload.quality ?? null,
  birthMonth: payload.birth_month ?? null,
  birthYear: payload.birth_year ?? null,
  siren: payload.siren ?? null,
  denomination: payload.denomination ?? null,
  nationality: payload.nationality ?? null,
});

const mapSireneNewBusiness = (payload: SireneNewBusinessResponse): SireneNewBusiness => ({
  siret: payload.siret,
  siren: payload.siren ?? null,
  nic: payload.nic ?? null,
  name: payload.name ?? null,
  nafCode: payload.naf_code ?? null,
  nafLabel: payload.naf_label ?? null,
  dateCreation: payload.date_creation ?? null,
  isIndividual: payload.is_individual,
  leaderName: payload.leader_name ?? null,
  denominationUniteLegale: payload.denomination_unite_legale ?? null,
  denominationUsuelleUniteLegale: payload.denomination_usuelle_unite_legale ?? null,
  denominationUsuelleEtablissement: payload.denomination_usuelle_etablissement ?? null,
  enseigne1: payload.enseigne1 ?? null,
  enseigne2: payload.enseigne2 ?? null,
  enseigne3: payload.enseigne3 ?? null,
  complementAdresse: payload.complement_adresse ?? null,
  numeroVoie: payload.numero_voie ?? null,
  indiceRepetition: payload.indice_repetition ?? null,
  typeVoie: payload.type_voie ?? null,
  libelleVoie: payload.libelle_voie ?? null,
  codePostal: payload.code_postal ?? null,
  libelleCommune: payload.libelle_commune ?? null,
  libelleCommuneEtranger: payload.libelle_commune_etranger ?? null,
  legalUnitName: payload.legal_unit_name ?? null,
  directors: (payload.directors ?? []).map(mapSireneDirector),
});

export const toolsApi = {
  fetchSireneNewBusinesses: async (
    payload: SireneNewBusinessesPayload,
  ): Promise<SireneNewBusinessesResult> => {
    const response = await request<SireneNewBusinessesResponse>("/admin/tools/sirene/new-establishments", {
      method: "POST",
      body: JSON.stringify({
        start_date: payload.startDate,
        end_date: payload.endDate ?? null,
        naf_codes: payload.nafCodes,
        limit: payload.limit ?? 100,
        department_codes: payload.departmentCodes && payload.departmentCodes.length > 0 ? payload.departmentCodes : null,
        enrich_annuaire: payload.enrichAnnuaire ?? false,
      }),
    });

    return {
      total: response.data.total,
      returned: response.data.returned,
      establishments: response.data.establishments.map(mapSireneNewBusiness),
    };
  },
  fetchAnnuaireDebug: async (siret: string): Promise<AnnuaireDebugResult> => {
    const response = await request<AnnuaireDebugResponse>(
      `/admin/tools/annuaire/debug?siret=${encodeURIComponent(siret)}`,
    );

    return {
      siret: response.data.siret,
      siren: response.data.siren,
      success: response.data.success,
      statusCode: response.data.status_code,
      durationMs: response.data.duration_ms,
      error: response.data.error ?? null,
      payload: response.data.payload ?? null,
    };
  },
};

File: app/api/schemas/tools.py
"""Schemas pour les outils admin (Sirene, etc.)."""
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field, field_validator, model_validator

from app.utils.naf import normalize_naf_code


class SireneNewBusinessesRequest(BaseModel):
    start_date: date = Field(description="Date de début (incluse) pour la création des établissements.")
    end_date: date | None = Field(
        default=None,
        description="Date de fin (incluse). Si absente, la date de début est utilisée.",
    )
    naf_codes: list[str] = Field(
        description="Liste des codes NAF à cibler (ex: 56.10A).",
        min_length=1,
    )
    limit: int = Field(
        default=100,
        ge=1,
        le=200,
        description="Nombre maximum d'établissements à retourner.",
    )
    department_codes: list[str] | None = Field(
        default=None,
        description="Codes départements à filtrer (ex: 75, 33).",
    )
    enrich_annuaire: bool = Field(
        default=False,
        description="Si activé, enrichit chaque résultat avec les dirigeants et le nom de l'unité légale via l'API Recherche Entreprises.",
    )

    @field_validator("naf_codes")
    @classmethod
    def validate_naf_codes(cls, value: list[str]) -> list[str]:
        normalized: list[str] = []
        seen: set[str] = set()
        for raw in value or []:
            candidate = (raw or "").strip().upper().replace(" ", "")
            normalized_code = normalize_naf_code(candidate)
            if not normalized_code:
                raise ValueError("Chaque code NAF doit contenir 4 chiffres suivis d'une lettre (ex: 56.10A).")
            if normalized_code in seen:
                continue
            seen.add(normalized_code)
            normalized.append(normalized_code)
            if len(normalized) > 25:
                raise ValueError("Maximum 25 codes NAF ciblés par requête.")
        if not normalized:
            raise ValueError("Au moins un code NAF est requis.")
        return normalized

    @model_validator(mode="after")
    def validate_dates(self) -> "SireneNewBusinessesRequest":
        if self.end_date and self.end_date < self.start_date:
            raise ValueError("La date de fin doit être postérieure ou égale à la date de début.")
        return self


class SireneNewBusinessDirectorOut(BaseModel):
    type_dirigeant: str
    first_names: str | None = None
    last_name: str | None = None
    quality: str | None = None
    birth_month: int | None = None
    birth_year: int | None = None
    siren: str | None = None
    denomination: str | None = None
    nationality: str | None = None


class SireneNewBusinessOut(BaseModel):
    siret: str
    siren: str | None = None
    nic: str | None = None
    name: str | None = None
    naf_code: str | None = None
    naf_label: str | None = None
    date_creation: date | None = None
    is_individual: bool
    leader_name: str | None = None
    denomination_unite_legale: str | None = None
    denomination_usuelle_unite_legale: str | None = None
    denomination_usuelle_etablissement: str | None = None
    enseigne1: str | None = None
    enseigne2: str | None = None
    enseigne3: str | None = None
    complement_adresse: str | None = None
    numero_voie: str | None = None
    indice_repetition: str | None = None
    type_voie: str | None = None
    libelle_voie: str | None = None
    code_postal: str | None = None
    libelle_commune: str | None = None
    libelle_commune_etranger: str | None = None
    legal_unit_name: str | None = None
    directors: list[SireneNewBusinessDirectorOut] = []


class AnnuaireDebugResponse(BaseModel):
    siret: str
    siren: str
    success: bool
    status_code: int | None = None
    duration_ms: float | None = None
    error: str | None = None
    payload: dict[str, object] | None = None


class SireneNewBusinessesResponse(BaseModel):
    total: int
    returned: int
    establishments: list[SireneNewBusinessOut]


__all__ = [
    "SireneNewBusinessDirectorOut",
    "SireneNewBusinessesRequest",
    "SireneNewBusinessesResponse",
    "SireneNewBusinessOut",
    "AnnuaireDebugResponse",
]
